File: Sample/ContextApi.CS/ContextApi.Tests/AssertException.cs
using System;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace ContextApi.Tests
{
    public static class AssertException
    {
        public static int Throws<TException>(Action action)
            where TException : Exception
        {
            if (action == null)
                throw new ArgumentNullException("action");

            try
            {
                action();
                string message =  "Should have thrown " + typeof(TException).Name;
                Assert.Fail(message);
            }
            catch (TException)
            {
                // OK
            }
            return 0;
        }
    }
}

File: Sample/ContextApi.VB/ContextApi.Tests/AssertException.vb
Imports System
Imports System.Linq
Imports Microsoft.VisualStudio.TestTools.UnitTesting

Public NotInheritable Class AssertException
    Private Sub New()
    End Sub

    Public Shared Sub Throws(Of TException As Exception)(ByVal _action As Action)
        If _action Is Nothing Then
            Throw New ArgumentNullException("action")
        End If

        Try
            _action()
            Dim message As String = "Should have thrown " & GetType(TException).Name
            Assert.Fail(message)
        Catch e1 As TException
            ' OK
        End Try
    End Sub
End Class
